activesupport/lib/active_support/event_reporter.rb

```diff
@@ -0,0 +1,488 @@
+  #
+  #   Rails.event.notify(UserCreatedEvent.new(id: 123, name: "John Doe"))
+  #   # Emits event:
+  #   #  {
+  #   #    name: "UserCreatedEvent",
+  #   #    payload: #<UserCreatedEvent:0x111>,
+  #   #    timestamp: 1738964843208679035,
+  #   #    source_location: { filepath: "path/to/file.rb", lineno: 123, label: "UserService#create" }
+  #   #  }
+  #
+  # These objects should represent schematized events and be serializable.
+  #
+  # ==== Default Encoders
+  #
+  # Rails provides default encoders for common serialization formats. Event objects and tags MUST
+  # implement +to_h+ to be serialized.
+  #
+  #   class JSONLogSubscriber
+  #     def emit(event)
+  #       # event = { name: "UserCreatedEvent", payload: { UserCreatedEvent: #<UserCreatedEvent:0x111> } }
+  #       json_data = ActiveSupport::EventReporter.encoder(:json).encode(event)
+  #       # => {
+  #       #      "name": "UserCreatedEvent",
+  #       #      "payload": {
+  #       #        "id": 123,
+  #       #        "name": "John Doe"
+  #       #      }
+  #       #    }
+  #       Rails.logger.info(json_data)
+  #     end
+  #   end
+  #
+  #   class MessagePackSubscriber
+  #     def emit(event)
+  #       msgpack_data = ActiveSupport::EventReporter.encoder(:msgpack).encode(event)
+  #       BatchExporter.export(msgpack_data)
+  #     end
+  #   end
+  #
+  # ==== Debug Events
+  #
+  # You can use the +debug+ method to report an event that will only be reported if the
+  # event reporter is in debug mode:
+  #
+  #   Rails.event.debug("my_debug_event", { foo: "bar" })
+  #
+  # ==== Tags
+  #
+  # To add additional context to an event, separate from the event payload, you can add
+  # tags via the +tagged+ method:
+  #
+  #   Rails.event.tagged("graphql") do
+  #     Rails.event.notify("user_created", { id: 123 })
+  #   end
+  #
+  #   # Emits event:
+  #   #  {
+  #   #    name: "user_created",
+  #   #    payload: { id: 123 },
+  #   #    tags: { graphql: true },
+  #   #    timestamp: 1738964843208679035,
+  #   #    source_location: { filepath: "path/to/file.rb", lineno: 123, label: "UserService#create" }
+  #   #  }
+  #
+  # ==== Context Store
+  #
+  # You may want to attach metadata to every event emitted by the reporter. While tags
+  # provide domain-specific context for a series of events, context is scoped to the job / request
+  # and should be used for metadata associated with the execution context.
+  # Context can be set via the +set_context+ method:
+  #
+  #   Rails.event.set_context(request_id: "abcd123", user_agent: TestAgent")
+  #   Rails.event.notify("user_created", { id: 123 })
+  #   # Emits event:
+  #   #  {
+  #   #    name: "user_created",
+  #   #    payload: { id: 123 },
+  #   #    context: { request_id: "abcd123", user_agent: TestAgent" },
+  #   #    timestamp: 1738964843208679035,
+  #   #    source_location: { filepath: "path/to/file.rb", lineno: 123, label: "UserService#create" }
+  #   #  }
```

Comment:
I can't tell if these code blocks need 4 spaces?
This one does look a little weird rendered in HTML, compared to the block after this for `CustomContextStore`:

https://shopify-ac-structured-events.rails-docs-preview.pages.dev/api/classes/ActiveSupport/EventReporter.html#class-ActiveSupport::EventReporter-label-Context+Store

Reply:
Oh thank you, yeah that definitely looks a little wonky!